plugins/processors/topk/topk.go

```diff
@@ -0,0 +1,402 @@
+package topk
+
+import (
+	"fmt"
+	"math"
+	"regexp"
+	"sort"
+	"strconv"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/processors"
+)
+
+type TopK struct {
+	Period            int
+	K                 int
+	GroupBy           []string `toml:"group_by"`
+	GroupByMetricName bool     `toml:"group_by_metric_name"`
+	Fields            []string
+	Aggregation       string
+	Bottomk           bool
+	SimpleTopk        bool   `toml:"simple_topk"`
+	DropNoGroup       bool   `toml:"drop_no_group"`
+	DropNonTop        bool   `toml:"drop_non_top"`
+	GroupByTag        string `toml:"group_by_tag"`
+	PositionField     string `toml:"position_field"`
+	AggregationField  string `toml:"aggregation_field"`
+
+	cache           map[string][]telegraf.Metric
+	metricRegex     *regexp.Regexp
+	tagsRegexes     map[string]*regexp.Regexp
+	lastAggregation time.Time
+}
+
+func NewTopK() TopK {
+	// Create object
+	topk := TopK{}
+
+	// Setup defaults
+	topk.Period = 10
+	topk.K = 10
+	topk.Fields = []string{"value"}
+	topk.Aggregation = "avg"
+	topk.GroupBy = []string{}
+	topk.GroupByMetricName = false
+	topk.GroupByTag = ""
+	topk.SimpleTopk = false
+	topk.DropNoGroup = true
+	topk.DropNonTop = true
+	topk.PositionField = ""
+	topk.AggregationField = ""
+
+	// Initialize cache
+	topk.Reset()
+
+	return topk
+}
+
+func NewTopKProcessor() telegraf.Processor {
+	topk := NewTopK()
+	return &topk
+}
+
+var sampleConfig = `
+[[processors.topk]]
```

Comment:
@danielnelson I've corrected this and several other issues. I've not pushed because in fixing this issue I changed lines tied to several open discussions, and I don't want them to show as updated when they are not yet resolved. When we come to a conclusion regarding the outstanding comments I will push my latest changes.